### ISSUE TITLE
onset/offset shift on onsets_and_frames result

### ISSUE THREAD

**Issue by TaegyunKwon** [tensorflow/magenta#1716]
Hi, i'm evaluating pretrained onset and frames transcription model.
Using the pretrained model, I transcribed MAESTRO dataset(test split) and saved the piano roll and midi. When I draw the histogram of note onset/offset errors compare to ground-truth midi, I found that estimated onsets are shifted around -16ms, while offsets are shifted +16ms (the center of the histogram of onset / offset errors was not zero-centered).
I suspect it is caused by sequence_to_pianoroll quantization, because start_frame is calculated with round down here https://github.com/tensorflow/magenta/blob/dc04ef3ff6a410559106fcf539a259153aae518d/magenta/music/sequences_lib.py#L1813
which might lead time shift of half frame(512/16000/2 = 16ms) when it converted back into time. 
I also checked hparam for min_frame_occupancy_for_label is set 0.0.

I also observed that mean onset F1 score improved (.9473 -> .9494) when I shifted prediction 16ms forward.

Is it right that the pretrained model is trained on shifted label? (+16ms on onset, -16ms on offset) if it is, should I shift the onset & offset to get correct result?

**Comment by TaegyunKwon** [comment]
I've checked with lasted version(master) today with one of wav file in MAESTRO set (MIDI-Unprocessed_03_R1_2008_01-04_ORIG_MID--AUDIO_03_R1_2008_wav--4.wav). Here's the piano roll of ground truth(above) and transcirption(below) .
Staccato notes at the beginning becomes longer (in both direction)

![onset_and_frames_rigoletto](https://user-images.githubusercontent.com/19758253/81641520-219c1280-945c-11ea-9287-8a6743473ff3.PNG)

Histogram of onset errors are also seems to be shifted.
![newplot (13)](https://user-images.githubusercontent.com/19758253/81641883-08e02c80-945d-11ea-829a-452fdce682be.png)

**Comment by cghawthorne** [comment]
Interesting! Maybe we should switch to using `np.round` (bankers' rounding) for both the onset and offset in `sequence_to_pianoroll` during training. Do you think that would fix the issues you're seeing?

**Comment by TaegyunKwon** [comment]
I think `min_frame_occupancy_for_label` parameter in `sequence_to_pianoroll` is determining the relation between sequence head and first frame in the piano roll.
https://github.com/tensorflow/magenta/blob/85ef5267513f62f4a40b01b2a1ee488f90f64a13/magenta/models/onsets_frames_transcription/data.py#L224
When `min_frame_occupancy_for_label == 0.5` might be the situation we are thinking of.
But I think it doesn't have to be like that, because label annotation can be determined arbitrary and  may have small or big effects on model performance. (have you already tested?)

I think we can compensate the effect of `min_frame_occupancy_for_label` when we decode piano roll to sequence. Maybe in `pianoroll_to_note_sequence`? https://github.com/tensorflow/magenta/blob/dc04ef3ff6a410559106fcf539a259153aae518d/magenta/music/sequences_lib.py#L1927

I'm not sure why offsets are delayed, while onsets are shifted forward. But maybe it happens with similar reason.

**Comment by cghawthorne** [comment]
I think using `min_frame_occupancy_for_label` would just shift the error distribution forward by a frame and using `np.around` would result in a more unbiased distribution of quantization errors.

@jesseengel do you have any thoughts here?
